src/query/storage/cache/redis_cache.go

```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
+		return results
+	}
+	for i, r := range responses {
+		// If the response is nil (not EmptyResult), then it doesn't exist
+		if len(r) == 0 {
+			results[i] = nil
+			cache.logger.Info("cache didn't get", zap.String("key", keys[i]))
+			continue
+		}
+		// Otherwise we did find it
+		res, err := resultDecode(r)
+		if err != nil {
+			cache.logger.Error("Redis decode error", zap.Error(err))
+		}
+		results[i] = res
+	}
+	return results
+}
+
+// For each entry in {entries} and associated value in {values},
+// sets entry -> value mapping in Redis
+// Array parameter used to handle multiple bucket requests in the future
+func (cache *RedisCache) Set(
+	entries []*storage.FetchQuery,
+	values []*storage.PromResult,
+) error {
+	var commands []radix.CmdAction
+	// Place holder variable to fit radix.Cmd() functions
+	var response string
+
+	for i := range entries {
+		// Skip over non-existent results
+		if entries[i] == nil {
+			continue
+		}
+		entry := keyEncode(entries[i])
+		value, err := resultEncode(values[i])
+		cache.logger.Info("Prom result size", zap.Int("length", len(value)), zap.Int("ts_len", len(values[i].PromResult.Timeseries)))
+		if err != nil {
+			cache.logger.Error("Redis encode error", zap.Error(err))
+			continue
+		}
+
+		commands = append(commands, radix.Cmd(&response, "SET", entry, value), radix.Cmd(&response, "EXPIRE", entry, ExpirationTime))
+	}
+
+	if err := cache.client.Do(radix.Pipeline(commands...)); err != nil {
+		cache.logger.Error("Failed to execute Redis set", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+// Function that checks whether the query data is in Redis, and returns that result if present
+// If it is not, it gets the result from M3DB
+//
+// Currently, this check is done by aligning queries to a minute so that queries asking for the
+// same timeseries over the same duration in the same minute are given the same result
+// (i.e. this is the key for Redis)
+func WindowGetOrFetch(
+	ctx context.Context,
+	st storage.Storage,
+	fetchOptions *storage.FetchOptions,
+	q *storage.FetchQuery,
+	cache *RedisCache,
+) (storage.PromResult, error) {
+	if cache == nil {
+		return st.FetchProm(ctx, q, fetchOptions)
+	}
+	query_range := int64(q.End.Sub(q.Start).Seconds())
+	// Align to the current minute
+	align_end := int64(q.End.Unix()/60) * 60
+	// Set start to aligned end minus the query range
+	align_start := align_end - query_range
+
+	align_q := &storage.FetchQuery{
+		TagMatchers: q.TagMatchers,
+		Start:       time.Unix(align_start, 0),
+		End:         time.Unix(align_end, 0),
+	}
+
+	res := cache.Get([]*storage.FetchQuery{align_q})
+	if res[0] == nil {
+		promRes, err := st.FetchProm(ctx, q, fetchOptions)
+		if err == nil {
+			cache.Set([]*storage.FetchQuery{align_q}, []*storage.PromResult{&promRes})
+		}
+
+		cache.logger.Info("cache miss", zap.String("key", keyEncode(align_q)), zap.Int("size", len(promRes.PromResult.Timeseries)))
```

Comment:
beside logging, can we also add metric ?

Reply:
Yeah, I have [another PR](https://github.com/andrewma2/m3/pull/4) for adding metrics that depends on this one, just split it into two for readability

Reply:
Update: Added metrics to this PR